text_file_to_ui/text_file_to_ui/UserParser.cs

```diff
@@ -12,25 +11,40 @@ public User UserParse(Reader a)
             a = userFileReader.FileOpen();
             if (a != null)
             {
-                b.firstName = a.firstNameString.Substring(4);
-                b.lastName = a.lastNameString.Substring(8);
-                b.yearOfBirth = a.yearOfBirthString.Substring(15);
-                b.city = a.cityString.Substring(13);
-                b.faculty = a.facultyString.Substring(9);
-                b.role = a.roleString.Substring(6);
-                if (b.role == "student")
+#if DEBUG
+                try
+#endif
                 {
-                    b.favoriteCourse = a.favoriteCourseString.Substring(17);
-                    b.department = " ";
+                    b.FirstName = a.FirstNameString.Substring(4);
+                    b.LastName = a.LastNameString.Substring(8);
+                    b.YearOfBirth = a.YearOfBirthString.Substring(15);
+                    b.City = a.CityString.Substring(13);
+                    b.Faculty = a.FacultyString.Substring(9);
+                    b.Role = a.RoleString.Substring(6);
+                    if (b.Role == "student")
+                    {
+                        b.FavoriteCourse = a.FavoriteCourseString.Substring(17);
+                        b.Department = " ";
+                    }
+                    else
+                    {
+                        b.Department = a.DepartmentString.Substring(8);
+                    }
+                    b.TimesRead = a.TimesRead.Substring(27);
+                    return b;
                 }
-                else
+#if DEBUG
+                catch
+#endif
                 {
-                    b.department = a.departmentString.Substring(8);
+                    MessageBox.Show("Unsupported file format.", "Error",
+                    MessageBoxButtons.OK, MessageBoxIcon.Error);
+                    return null;
```

Comment:
Ovo je dobro. Jedino, tu postoji jedna mini-arhitekturalna pogreška. Naime, ako već ova metoda vraća null, onda nije njezina odgovornost i stvoriti MessageBox, već neka njega stvara netko tko bi imao odgovornost obavijestiti korisnika o pogrešci.
Konkretno, budući da je MessageBox usko vezan uz grafičko sučelje, na formi treba stvoriti MessageBox ako je User bio null (kod provjere u formi `if (a != null)` dodati `else` i ovaj MessageBox).

Ima jedan vrlo bitan razlog zašto tako razmišljati. Recimo da sutra odlučimo isti ovaj program napraviti za Mac/Linux/konzolu. Ako ovako nonšalantno "pucamo" MessageBoxove (usko vezane uz OS Windows) u klasama poslovne logike (ova klasa stvara iz `readera` pravi objekt `User`), morat ćemo za te druge platforme prolaziti kroz cijelu bazu kôda i tražiti sva mjesta gdje stvaramo MessageBox.
Međutim, ako su nam svi MessageBoxovi samo na onome što je ionako usko vezano uz Windows (a to su Windows forme), onda smo sigurni da je ovaj kôd u klasama potpuno cross-platform i da su jedine ovisnosti na Windowse forme koje ionako idemo mijenjati za neke Mac/Linux prozore ili konzolnu aplikaciju. Tada je kôd **ponovno iskoristiv** što nam je uvijek jako bitno.

Zbog toga težimo forme (sloj korisničkog sučelja) držati minimalnim što se kôda poslovne logike tiče (zato stvaramo mnogo klasa), ali opet težimo na njega svaliti sve što ima veze s grafičkim sučeljem, poput skočnih MessageBoxova.